### DISCUSSION TITLE
Rect intersection inplace operation

### DISCUSSION THREAD

**Discussion by imntl** [pymupdf/PyMuPDF#2489] [Q&A] [answered]
Hello,

I have a question regarding the intersect operation of the `rect` object and could not find any information in issues or other discussions. As found in the [documentation](https://pymupdf.readthedocs.io/en/latest/rect.html#Rect.intersect):

> The intersection (common rectangular area, largest rectangle contained in both) of the current rectangle and r is calculated and **replaces the current** rectangle.

My question would be, why it is important that the intersection _replaces the current_, as sometimes it might be handy to have both objects as well as the intersection after the calculation. Adding a deep copy of my first object seems unnecessarily complex as the resulting intersection will always be given by the `return`. I just want to understand if that is intentionally or if it would be possible to add an optional `inplace=True` option to `intersect`, which then can be used to not change the `rect` object inplace. Any enlightenment will be appreciated.

## Example
```python
import fitz

rect1 = fitz.Rect(0,0,1,1)
rect2 = fitz.Rect(2,0,3,1)

print(f"Rect1: {rect1}")
print(f"Rect2: {rect2}")

inter = rect1.intersect(rect2)
print(f"Itersection: {inter}")
print(f"Rect1: {rect1}")
print(f"Rect2: {rect2}")
```

```bash
Rect1: Rect(0.0, 0.0, 1.0, 1.0)
Rect2: Rect(2.0, 0.0, 3.0, 1.0)
Itersection: Rect(2.0, 0.0, 1.0, 1.0)
Rect1: Rect(2.0, 0.0, 1.0, 1.0)
Rect2: Rect(2.0, 0.0, 3.0, 1.0)
```

You can see in the output, that `rect1` will get replaced by the intersection. This reflects what is written in the documentation, but deletes information from `rect1` to have the information of the intersection twice.

### System Info (for reproduction purposes, if needed):
3.11.4 (main, Jun  7 2023, 10:13:09) [GCC 12.2.0] 
 linux 
 
PyMuPDF 1.22.5: Python bindings for the MuPDF 1.22.2 library.
Version date: 2023-06-21 00:00:01.
Built for Python 3.11 on linux (64-bit).

**Comment by JorjMcKie** [comment, marked as answer]
You in fact have all freedom you need - just use the notation that suits you best.

1. This `r1.intersect(r2)` logic is the same as implemented in the base library.
2. There exists algebra notation for all geometry objects which lets you say `r = r1 & r2`. Here, the two original rectangles remain unchanged.
3. You can also write `r1 = r1 & r2` which obviously replaces r1 with result and is exactly the same as `r1 &= r2` and corresponds to `r1.intersect(r2)`.
